File: solutions/Two Pointers/Easy/125.valid-palindrome.py
#
# @lc app=leetcode id=125 lang=python3
#
# [125] Valid Palindrome
#

# @lc code=start
class Solution:
    def isPalindrome(self, s: str) -> bool:
        start, end = 0, len(s)-1

        # my code:
        # while start < end:
        #     if s[start].isalnum():
        #         if s[end].isalnum():
        #             if s[start].lower() == s[end].lower():
        #                 end -= 1
        #                 start +=1
        #             else: return False
        #         else: end -= 1
        #     else: start += 1

        # return True

        # gpt's code:
        while start < end:
            while start < end and not s[start].isalnum():
                start += 1
            
            while start < end and not s[end].isalnum():
                end -= 1

            if s[start].lower() != s[end].lower():
                return False

            start += 1
            end -= 1
        
        return True
            
# @lc code=end

File: 125.valid-palindrome.java
/*
 * @lc app=leetcode id=125 lang=java
 *
 * [125] Valid Palindrome
 * goal: palindrome -(Character.toLowerCase(), Character.isLetterOrDigit())
 * solution: two ptrs
 */

// @lc code=start
class Solution {
    public boolean isPalindrome(String s) {

        // s = s.replaceAll("\\s", ""); //must re-assign to the string

        int start = 0;

        int end = s.length()-1;

        while (start < end) {
            
            char left = s.charAt(start);
            char right = s.charAt(end);

            if (!Character.isLetterOrDigit(left)) {
                start++;
                continue; // Skip and restart loop
            }

            if (!Character.isLetterOrDigit(right)) {
                end--;
                continue;
            }

            if (Character.toLowerCase(left) != Character.toLowerCase(right)) return false;

            start++;
            end--;
        }

        return true;
        
    }
}
// @lc code=end

